include/libsemigroups/cache.hpp

```diff
@@ -0,0 +1,204 @@
+
+    // Declarations
+    template <typename T, typename = void>
+    class Cache;
+
+    template <typename T, typename = void>
+    class CacheGuard;
+
+    // Helpers
+    template <typename T>
+    using is_pointer_t =
+        typename std::enable_if<std::is_pointer<T>::value>::type;
+
+    template <typename T>
+    using is_non_pointer_t =
+        typename std::enable_if<!std::is_pointer<T>::value>::type;
+
+    // Cache for non-pointer types, these are intended to be used with the
+    // T = BruidhinnTraits::internal_value_type, and so the non-pointer objects
+    // should be small and easy to copy. There are only versions for pointer
+    // and non-pointer types as BruidhinnTraits::internal_value_type is always
+    // one of these two. It's your responsibility to ensure that the
+    // non-pointer type is actually small and easy to copy.
+    template <typename T>
+    class Cache<T, is_non_pointer_t<T>> final {
+      static_assert(std::is_default_constructible<T>::value,
+                    "Cache<T> requires T to be default-constructible");
+
+     public:
+      Cache()  = default;
+      ~Cache() = default;
+
+      // Deleted other constructors to avoid unintentional copying
+      Cache(Cache const&) = delete;
+      Cache(Cache&&)      = delete;
+      Cache& operator=(Cache const&) = delete;
+      Cache& operator=(Cache&&) = delete;
+
+      T acquire() {
+        return T();
+      }
+
+      void release(T&) {}
+
+      void push(T const&, size_t = 1) {}
+    };
+
+    template <typename T>
+    class Cache<T, is_pointer_t<T>> final {
+     public:
+      // Not noexcept because default constructors of, say, std::list isn't
+      Cache() = default;
+      ~Cache() {
+        while (!_acquirable.empty()) {
+          delete _acquirable.top();
+          _acquirable.pop();
+        }
+      }
+
+      // Deleted other constructors to avoid unintentional copying
+      Cache(Cache const&) = delete;
+      Cache(Cache&&)      = delete;
+      Cache& operator=(Cache const&) = delete;
+      Cache& operator=(Cache&&) = delete;
+
+      // Not noexcept because it can throw
+      T acquire() {
+        if (_acquirable.empty()) {
+          LIBSEMIGROUPS_EXCEPTION(
+              "attempted to acquire an object, but none are acquirable");
+        }
+        T ptr = _acquirable.top();
+        _acquirable.pop();
+        _acquired.push_back(ptr);
+        _map.emplace(ptr, --_acquired.end());
+        return ptr;
+      }
+
+      // Not noexcept because it can throw
+      void release(T ptr) {
+        auto it = _map.find(ptr);
+        if (it == _map.end()) {
+          LIBSEMIGROUPS_EXCEPTION("attempted to release an object which is "
+                                  "not owned by this cache");
+        }
+        _acquired.erase(it->second);
+        _map.erase(it);
+        _acquirable.push(ptr);
+      }
+
+      // noexcept because std::list<T>::size is noexcept.
+      size_t acquired() const noexcept {
```

Comment:
`acquired` and `acquirable` are only available in this version and are used for testing only. I'm not sure they are really necessary.